### ISSUE TITLE
Methods on sftp.Request mix pointer and value receivers

### ISSUE THREAD

**Issue by davecheney** [pkg/sftp#192]
gopl recommends that all methods be on the same type of receiver; either pointer or value, and give the gynmastics that newRequest goes through to avoid copying a mutex and the state object each time Request is called as a value receiver or passed by value, it's clear the type of Requets' receiver should be *Request.

**Comment by eikenb** [comment]
I've read before about the idea of all methods keeping a uniform pointer or value receiver, but not mixing the 2. I've also read it best to use values whenever possible as they are more stack friendly. In practice I tend toward the latter and default to values, using pointers when necessary. I haven't picked up that book yet but I respect the authors, can you summarize their reasoning?

**Comment by eikenb** [comment]
Forgot to mention, I also like how using pointers vs. values can be used to make the code more self documenting. By using a pointer receiver you are denoting that method changes the state of the struct, vs. using a value which means the opposite.

**Comment by eikenb** [comment]
The joke is on me. I just got bit by this. By mixing the 2 you can end up with a pointer to a copy of the struct pretty easily. So the lesson is that when you need pointer receivers it's best to have them everywhere or you need to track the path of every pointer receiver call to make sure all their calling methods on the same struct are also pointer receivers... better to just make them all receivers and avoid these subtle bugs.

I'll be merging your PR.

**Comment by davecheney** [comment]
Yup, basically that's the punchline.

**Comment by eikenb** [comment]
Had it in my head that your PR changed all the methods to use pointer receivers. Re-opening this until I've actually addressed it.
